File: examples/account/signing.rs
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! cargo run --example account_signing

use std::path::PathBuf;

use identity_iota::account::Account;
use identity_iota::account::IdentitySetup;
use identity_iota::account::MethodContent;
use identity_iota::account::Result;
use identity_iota::account_storage::Stronghold;
use identity_iota::core::json;
use identity_iota::core::FromJson;
use identity_iota::core::Url;
use identity_iota::credential::Credential;
use identity_iota::credential::Subject;
use identity_iota::crypto::KeyPair;
use identity_iota::crypto::ProofOptions;
use identity_iota::did::verifiable::VerifierOptions;
use identity_iota::did::DID;
use identity_iota::iota::ExplorerUrl;
use identity_iota::iota::ResolvedIotaDocument;
use identity_iota::iota_core::IotaDID;
use identity_iota::prelude::*;

#[tokio::main]
async fn main() -> Result<()> {
  pretty_env_logger::init();

  // ===========================================================================
  // Create Identity - Similar to create_did example
  // ===========================================================================

  // Stronghold settings
  let stronghold_path: PathBuf = "./example-strong.hodl".into();
  let password: String = "my-password".to_owned();
  let stronghold: Stronghold = Stronghold::new(&stronghold_path, password, None).await?;

  // Create a new Account with stronghold storage.
  let mut account: Account = Account::builder()
    .storage(stronghold)
    .create_identity(IdentitySetup::default())
    .await?;

  // ===========================================================================
  // Signing Example
  // ===========================================================================

  // Add a new Ed25519 Verification Method to the identity
  account
    .update_identity()
    .create_method()
    .content(MethodContent::GenerateEd25519)
    .fragment("key-1")
    .apply()
    .await?;

  // Create a subject DID for the recipient of a `UniversityDegree` credential.
  let subject_key: KeyPair = KeyPair::new(KeyType::Ed25519)?;
  let subject_did: IotaDID = IotaDID::new(subject_key.public().as_ref())?;

  // Create the actual Verifiable Credential subject.
  let subject: Subject = Subject::from_json_value(json!({
    "id": subject_did.as_str(),
    "degree": {
      "type": "BachelorDegree",
      "name": "Bachelor of Science and Arts"
    }
  }))?;

  // Issue an unsigned Credential...
  let mut credential: Credential = Credential::builder(Default::default())
    .issuer(Url::parse(account.did().as_str())?)
    .type_("UniversityDegreeCredential")
    .subject(subject)
    .build()?;

  // ...and sign the Credential with the previously created Verification Method
  account.sign("key-1", &mut credential, ProofOptions::default()).await?;

  println!("[Example] Local Credential = {:#}", credential);

  // Fetch the DID Document from the Tangle
  //
  // This is an optional step to ensure DID Document consistency.
  let resolved: ResolvedIotaDocument = account.resolve_identity().await?;

  // Retrieve the DID from the newly created identity.
  let iota_did: &IotaDID = account.did();

  // Prints the Identity Resolver Explorer URL.
  // The entire history can be observed on this page by clicking "Loading History".
  let explorer: &ExplorerUrl = ExplorerUrl::mainnet();
  println!(
    "[Example] Explore the DID Document = {}",
    explorer.resolver_url(iota_did)?
  );

  // Ensure the resolved DID Document can verify the credential signature
  let verified: bool = resolved
    .document
    .verify_data(&credential, &VerifierOptions::default())
    .is_ok();

  println!("[Example] Credential Verified = {}", verified);

  Ok(())
}

File: bindings/wasm/examples-account/src/signing.ts
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    AccountBuilder,
    Credential,
    DID,
    ExplorerUrl,
    KeyPair,
    KeyType,
    MethodContent,
    ProofOptions,
    Storage,
    VerifierOptions
} from './../../node';

/**
 * This example demonstrates how to issue and sign Verifiable Credentials using the account.
 */
async function signing(storage?: Storage) {

    // ===========================================================================
    // Create Identity - Similar to create_did example
    // ===========================================================================

    // The creation step generates a keypair, builds an identity
    // and publishes it to the IOTA mainnet.
    let builder = new AccountBuilder({
        storage,
    });
    let account = await builder.createIdentity();

    // ===========================================================================
    // Signing Example
    // ===========================================================================

    // Add a new Ed25519 Verification Method to the identity for signing issued verifiable credentials.
    await account.createMethod({
        content: MethodContent.GenerateEd25519(),
        fragment: "key_1"
    })

    // Prepare a credential subject indicating the degree earned by Alice, linked to their DID.
    const subject = {
        id: "did:iota:3TT7QsmESw1dcboV2oVTAuJbbxaVYAThexZFct2z5Q2d",
        name: "Alice",
        degree: {
            type: "BachelorDegree",
            name: "Bachelor of Science and Arts"
        }
    };

    // Issue an unsigned Credential...
    const unsignedVc = new Credential({
        issuer: account.did(),
        type: "UniversityDegreeCredential",
        credentialSubject: subject,
    });

    // ...and sign the Credential with the previously created Verification Method.
    // Note: Different methods are available for different data types,
    // use the Method `createSignedData` to sign arbitrary data.
    let signedVc = await account.createSignedCredential("key_1", unsignedVc, ProofOptions.default());

    console.log("[Example] Local Credential", signedVc);

    // Fetch the DID Document from the Tangle.
    //
    // This is an optional step to ensure DID Document consistency.
    let resolved = await account.resolveIdentity();

    // Retrieve the DID from the newly created identity.
    let did = account.did().toString();

    // Print the Explorer URL for the DID.
    console.log(`Explorer Url:`, ExplorerUrl.mainnet().resolverUrl(did));

    // Ensure the resolved DID Document can verify the credential signature.
    let verified = resolved.intoDocument().verifyData(signedVc, VerifierOptions.default());

    console.log("[Example] Credential Verified = ", verified);
}

export {signing};
